src/common/guards/DocuNinjaGuard.tsx

```diff
@@ -0,0 +1,129 @@
+import { useCurrentCompanyUser } from '$app/common/hooks/useCurrentCompanyUser';
+import { useCurrentUser } from '$app/common/hooks/useCurrentUser';
+import { CompanyUser } from '$app/common/interfaces/company-user';
+import { User } from '$app/common/interfaces/user';
+import { SettingsLevel } from '../stores/slices/settings';
+import { useActiveSettingsDetails } from '../hooks/useActiveSettingsDetails';
+import { DocuNinjaData } from '$app/common/atoms/docuninja';
+import { useDocuNinjaData } from '$app/common/hooks/useDocuNinjaData';
+
+// Re-export DocuNinjaData for backward compatibility
+export type { DocuNinjaData };
+
+export type DocuNinjaGuard = (ctx: DocuNinjaContext) => Promise<boolean>;
+
+export interface DocuNinjaContext {
+  queryClient: QueryClient;
+  params: Params<string>;
+  user: User | undefined;
+  companyUser: CompanyUser | undefined;
+  settingsLevel: SettingsLevel;
+  docuData?: DocuNinjaData;
+}
+
+// DocuNinjaData is now imported from atoms
+
+export interface DocuNinjaPermission {
+  model: string;
+  permission_id: number;
+  action?: string;
+}
+
+export interface DocuNinjaGuardProps {
+  guards: DocuNinjaGuard[];
+  component: React.ReactElement;
+  type?: 'page' | 'subPage' | 'component';
+  docuData?: DocuNinjaData;
+}
+
+export function useDocuNinjaGuardContext(docuData?: DocuNinjaData) {
+  const queryClient = useQueryClient();
+  const params = useParams();
+  const user = useCurrentUser();
+  const companyUser = useCurrentCompanyUser();
+  const activeSettings = useActiveSettingsDetails();
+  
+  // Get DocuNinja data from unified atoms (NO QUERY!)
+  const unifiedDocuData = useDocuNinjaData();
+
+  // Use provided docuData, unified data, or null if not available
+  const globalDocuData = docuData || unifiedDocuData || null;
+
+  return {
+    queryClient,
+    params,
+    user,
+    companyUser,
+    settingsLevel: activeSettings.level,
+    docuData: globalDocuData,
+  };
+}
+
+enum DocuNinjaState {
+  Authorized = 'authorized',
+  Loading = 'loading',
+  Unauthorized = 'unauthorized',
+}
+
+export function DocuNinjaGuard({ guards, component, type = 'page', docuData }: DocuNinjaGuardProps) {
+  const [state, setState] = useState<DocuNinjaState>(DocuNinjaState.Loading);
+  const { companyUser, queryClient, params, user, settingsLevel, docuData: contextDocuData } =
+    useDocuNinjaGuardContext(docuData);
+
+  useEffect(() => {
+    const promises = guards.map((guard) =>
+      guard({ companyUser, queryClient, params, user, settingsLevel, docuData: contextDocuData || undefined })
+    );
+
+    Promise.all(promises)
+      .then((values) => {
+        values.includes(false)
+          ? setState(DocuNinjaState.Unauthorized)
+          : setState(DocuNinjaState.Authorized);
+      })
+      .catch((error) => {
+        setState(DocuNinjaState.Unauthorized);
+      });
+  }, [guards, companyUser, queryClient, params, user, settingsLevel, contextDocuData]);
```

Comment:
What's the reason behind reimplementing guards logic? Isn't current one enough with couple more guards?